sympy/polys/rings.py

```diff
@@ -1103,101 +973,15 @@ def __rsub__(p1, n):
         >>> p = x + y
         >>> 4 - p
         -x - y + 4
-
         """
         ring = p1.ring
         try:
             n = ring.domain_new(n)
         except CoercionFailed:
             return NotImplemented
         else:
-            p = ring.zero
-            for expv in p1:
-                p[expv] = -p1[expv]
-            p += n
-            # p._check()
-            return p
-
-    def __mul__(p1, p2):
-        """Multiply two polynomials.
-
-        Examples
-        ========
-
-        >>> from sympy.polys.domains import QQ
-        >>> from sympy.polys.rings import ring
-
-        >>> _, x, y = ring('x, y', QQ)
-        >>> p1 = x + y
-        >>> p2 = x - y
-        >>> p1*p2
-        x**2 - y**2
-
-        """
-        ring = p1.ring
-        p = ring.zero
-        if not p1 or not p2:
-            return p
-        elif ring.is_element(p2):
-            get = p.get
-            zero = ring.domain.zero
-            monomial_mul = ring.monomial_mul
-            p2it = list(p2.items())
-            for exp1, v1 in p1.items():
-                for exp2, v2 in p2it:
-                    exp = monomial_mul(exp1, exp2)
-                    p[exp] = get(exp, zero) + v1*v2
-            p.strip_zero()
-            # p._check()
-            return p
-        elif isinstance(p2, PolyElement):
-            if isinstance(ring.domain, PolynomialRing) and ring.domain.ring == p2.ring:
-                pass
-            elif isinstance(p2.ring.domain, PolynomialRing) and p2.ring.domain.ring == ring:
-                return p2.__rmul__(p1)
-            else:
-                return NotImplemented
-
-        try:
-            p2 = ring.domain_new(p2)
-        except CoercionFailed:
-            return NotImplemented
-        else:
-            for exp1, v1 in p1.items():
-                v = v1*p2
-                if v:
-                    p[exp1] = v
-            # p._check()
-            return p
-
-    def __rmul__(p1, p2):
-        """p2 * p1 with p2 in the coefficient domain of p1.
-
-        Examples
-        ========
-
-        >>> from sympy.polys.domains import ZZ
-        >>> from sympy.polys.rings import ring
-
-        >>> _, x, y = ring('x, y', ZZ)
-        >>> p = x + y
-        >>> 4 * p
-        4*x + 4*y
-
-        """
-        p = p1.ring.zero
-        if not p2:
-            return p
-        try:
-            p2 = p.ring.domain_new(p2)
-        except CoercionFailed:
-            return NotImplemented
-        else:
-            for exp1, v1 in p1.items():
-                v = p2*v1
-                if v:
-                    p[exp1] = v
-            return p
+            p = p1._negate()
+            return p._add_ground(n)
```

Comment:
It would be better to have an `_rsub_ground` method for this. It can be done in one step with python-flint.

Reply:
In python-flint this is just `n - p`:
```python
In [2]: ctx = flint.fmpz_mpoly_ctx.get(['x', 'y'])

In [3]: x, y = ctx.gens()

In [4]: c = flint.fmpz(2)

In [5]: c - x
Out[5]: -x + 2
```
We don't need to create an intermediate negated polynomial.